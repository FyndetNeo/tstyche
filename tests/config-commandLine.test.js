import { afterEach, describe, expect, test } from "@jest/globals";
import { clearFixture, writeFixture } from "./__utils__/fixtureFactory.js";
import { normalizeOutput } from "./__utils__/normalizeOutput.js";
import { spawnTyche } from "./__utils__/spawnTyche.js";

const isStringTestText = `import { expect, test } from "tstyche";
test("is string?", () => {
  expect<string>().type.toBeString();
});
`;

const isNumberTestText = `import { expect, test } from "tstyche";
test("is number?", () => {
  expect<number>().type.toBeNumber();
});
`;

const fixture = "config-commandLine";

afterEach(async () => {
  await clearFixture(fixture);
});

describe("'tstyche' command", () => {
  test("when called without arguments, selects all matching files", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isNumber.test.ts"]: isNumberTestText,
      ["__typetests__/isString.test.ts"]: isStringTestText,
      ["feature/__tests__/isNumber.tst.ts"]: isNumberTestText,
      ["feature/__tests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when single search string is provided, selects matching files", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isNumber.test.ts"]: isNumberTestText,
      ["__typetests__/isString.test.ts"]: isStringTestText,
      ["feature/__tests__/isNumber.tst.ts"]: isNumberTestText,
      ["feature/__tests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["number"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when multiple search strings are provided, selects matching files", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isNumber.test.ts"]: isNumberTestText,
      ["__typetests__/isString.test.ts"]: isStringTestText,
      ["feature/__tests__/isNumber.tst.ts"]: isNumberTestText,
      ["feature/__tests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["string", "feature"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("when relative search string is provided, selects matching files", async () => {
    await writeFixture(fixture, {
      ["__typetests__/isNumber.test.ts"]: isNumberTestText,
      ["__typetests__/isString.test.ts"]: isStringTestText,
      ["feature/__tests__/isNumber.tst.ts"]: isNumberTestText,
      ["feature/__tests__/isString.tst.ts"]: isStringTestText,
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["./feature/__tests__/isNumber"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });

  test("only the files matched by the 'testFileMatch' patterns are selected", async () => {
    const config = {
      testFileMatch: ["**/feature/__tests__/**.*"],
    };
    const tsconfig = { extends: "../../../tsconfig.json", include: ["**/*"] };

    await writeFixture(fixture, {
      ["__typetests__/isNumber.tst.ts"]: isNumberTestText,
      ["__typetests__/isString.tst.ts"]: isStringTestText,
      ["feature/__tests__/__snapshot__/isNumber.test.ts.snap"]: "isNumberSnap",
      ["feature/__tests__/__snapshot__/isString.test.ts.snap"]: "isStringSnap",
      ["feature/__tests__/isNumber.test.ts"]: isNumberTestText,
      ["feature/__tests__/isString.test.ts"]: isStringTestText,
      ["feature/__tests__/tsconfig.json"]: JSON.stringify(tsconfig, null, 2),
      ["tstyche.config.json"]: JSON.stringify(config, null, 2),
    });

    const { exitCode, stderr, stdout } = await spawnTyche(fixture, ["number"]);

    expect(normalizeOutput(stdout)).toMatchSnapshot("stdout");
    expect(stderr).toBe("");

    expect(exitCode).toBe(0);
  });
});
