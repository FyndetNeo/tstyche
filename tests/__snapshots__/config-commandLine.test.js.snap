// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'tstyche' command only the files matched by the 'testFileMatch' patterns are selected: stdout 1`] = `
"uses TypeScript <<version>> with ./feature/__tests__/tsconfig.json

pass ./feature/__tests__/isNumber.test.ts
  + is number?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran test files matching 'number'.
"
`;

exports[`'tstyche' command when called without arguments, selects all matching files: stdout 1`] = `
"uses TypeScript <<version>>

pass ./feature/__tests__/isNumber.tst.ts
pass ./feature/__tests__/isString.tst.ts
pass ./__typetests__/isNumber.test.ts
pass ./__typetests__/isString.test.ts

Targets:    1 passed, 1 total
Test files: 4 passed, 4 total
Tests:      4 passed, 4 total
Assertions: 4 passed, 4 total
Duration:   <<timestamp>>

Ran all test files.
"
`;

exports[`'tstyche' command when multiple search strings are provided, selects matching files: stdout 1`] = `
"uses TypeScript <<version>>

pass ./feature/__tests__/isNumber.tst.ts
pass ./feature/__tests__/isString.tst.ts
pass ./__typetests__/isString.test.ts

Targets:    1 passed, 1 total
Test files: 3 passed, 3 total
Tests:      3 passed, 3 total
Assertions: 3 passed, 3 total
Duration:   <<timestamp>>

Ran test files matching 'string' or 'feature'.
"
`;

exports[`'tstyche' command when relative search string is provided, selects matching files: stdout 1`] = `
"uses TypeScript <<version>>

pass ./feature/__tests__/isNumber.tst.ts
  + is number?

Targets:    1 passed, 1 total
Test files: 1 passed, 1 total
Tests:      1 passed, 1 total
Assertions: 1 passed, 1 total
Duration:   <<timestamp>>

Ran test files matching './feature/__tests__/isNumber'.
"
`;

exports[`'tstyche' command when single search string is provided, selects matching files: stdout 1`] = `
"uses TypeScript <<version>>

pass ./feature/__tests__/isNumber.tst.ts
pass ./__typetests__/isNumber.test.ts

Targets:    1 passed, 1 total
Test files: 2 passed, 2 total
Tests:      2 passed, 2 total
Assertions: 2 passed, 2 total
Duration:   <<timestamp>>

Ran test files matching 'number'.
"
`;
